Porting BottomSheet accessibility from v2 to v4

Mahmoud-SK opened the pull request:
## Motivation

I'm picking up this [task](https://github.com/gorhom/react-native-bottom-sheet/pull/693) of porting accessibility features from V2 to V4 as per [#548](https://github.com/gorhom/react-native-bottom-sheet/issues/548)

Mahmoud-SK committed:
Added accessibility to the bottomSheet component

Mahmoud-SK committed:
Accessibility changes - BottomSheetBackdrop

Mahmoud-SK committed:
Accessibility changes - BottomSheetHandle

gorhom commented on src/components/bottomSheet/types.d.ts:
```diff
@@ -295,6 +299,21 @@ export interface BottomSheetProps
+   * Callback on snap position change that announces the returned string on screen readers.
+   * @type string | ((positionInScreen: string) => string)
+   */
+  accessibilityPositionChangeAnnouncement?:
```
these two props could be handled with `onChange` by the user outside of the library, i don't think we need to handle it internally

Mahmoud-SK replied:
I've went ahead and removed these two props.

jenskuhrjorgensen commented on src/components/bottomSheetBackdrop/BottomSheetBackdrop.tsx:
```diff
@@ -119,12 +128,11 @@ const BottomSheetBackdropComponent = ({
+        accessibilityRole={_providedAccessibilityRole ?? undefined}
+        accessibilityLabel={_providedAccessibilityLabel ?? undefined}
+        accessibilityHint={_providedAccessiblityHint ?? undefined}
+        {...rest}
```
@Mahmoud-SK why is the spread of the `rest` props needed?

Mahmoud-SK replied:
It's needed to pass the rest of the properties to the Animated.View - in this case it's only animatedPosition-prop.

jenskuhrjorgensen commented on src/components/bottomSheetBackdrop/BottomSheetBackdrop.tsx:
```diff
@@ -119,12 +128,11 @@ const BottomSheetBackdropComponent = ({
+        accessible={_providedAccessible ?? undefined}
+        accessibilityRole={_providedAccessibilityRole ?? undefined}
+        accessibilityLabel={_providedAccessibilityLabel ?? undefined}
+        accessibilityHint={_providedAccessiblityHint ?? undefined}
```
@Mahmoud-SK can we maybe do this in order to not break the existing behaviour?
```
accessibilityHint ={_providedAccessiblityHint  ? _providedAccessibilityHint  : `Tap to ${
          typeof pressBehavior === 'string' ? pressBehavior : 'move'
        } the Bottom Sheet`}
```

Mahmoud-SK replied:
Done. :)

jenskuhrjorgensen commented on src/components/bottomSheet/BottomSheet.tsx:
```diff
@@ -323,7 +341,7 @@ const BottomSheetComponent = forwardRef<BottomSheet, BottomSheetProps>(
         0,
         animatedContainerHeight.value -
-          (animatedSheetHeight.value + keyboardHeightInContainer)
+        (animatedSheetHeight.value + keyboardHeightInContainer)
```
I think this was correctly indented before - can you revert this change?

Mahmoud-SK replied:
It looks correctly indented locally, so I'm not completely sure why it shows it like this.

jenskuhrjorgensen commented on src/components/bottomSheet/BottomSheet.tsx:
```diff
@@ -473,11 +491,11 @@ const BottomSheetComponent = forwardRef<BottomSheet, BottomSheetProps>(
+          adjustedSnapPoints,
+          adjustedSnapPointsIndexes,
+          Extrapolate.CLAMP
+        )
```
I think this was correctly indented before - can you revert this change?

Mahmoud-SK replied:
Same as the previous comment, locally it's indented.

jenskuhrjorgensen commented on src/components/bottomSheet/BottomSheet.tsx:
```diff
@@ -1393,9 +1444,9 @@ const BottomSheetComponent = forwardRef<BottomSheet, BottomSheetProps>(
-            )
+            _keyboardHeight -
+            Math.abs(bottomInset - animatedContainerOffset.value.bottom)
+          )
```
Wasn't this also correctly indented before?

Mahmoud-SK replied:
Same as the previous comment, locally it's indented.

jenskuhrjorgensen commented on src/components/bottomSheet/BottomSheet.tsx:
```diff
@@ -1635,6 +1686,10 @@ const BottomSheetComponent = forwardRef<BottomSheet, BottomSheetProps>(
+                  accessible={_providedAccessible ?? undefined}
+                  accessibilityRole={_providedAccessibilityRole ?? undefined}
+                  accessibilityLabel={_providedAccessibilityLabel ?? undefined}
+                  {...rest}
```
Why is the spread of the `rest` props needed?

jenskuhrjorgensen commented on src/components/bottomSheet/BottomSheet.tsx:
```diff
@@ -159,6 +165,18 @@ const BottomSheetComponent = forwardRef<BottomSheet, BottomSheetProps>(
+      _providedEnableAccessibilityChangeAnnouncement = DEFAULT_ENABLE_ACCESSIBILITY_CHANGE_ANNOUNCEMENT,
+      accessibilityPositionChangeAnnouncement:
+      _providedAccessibilityPositionChangeAnnouncement = DEFAULT_ACCESSIBILITY_POSITION_CHANGE_ANNOUNCEMENT,
+      ...rest
```
Why is the spread of the `rest` props needed?

jenskuhrjorgensen commented on src/components/bottomSheetBackdrop/BottomSheetBackdrop.tsx:
```diff
@@ -33,6 +37,11 @@ const BottomSheetBackdropComponent = ({
+  accessibilityRole: _providedAccessibilityRole = DEFAULT_ACCESSIBILITY_ROLE,
+  accessibilityLabel: _providedAccessibilityLabel = DEFAULT_ACCESSIBILITY_LABEL,
+  accessibilityHint: _providedAccessiblityHint = DEFAULT_ACCESSIBILITY_HINT,
+  ...rest
```
Why is the spread of the `rest` props needed?

Mahmoud-SK replied:
It currently contains the animatedPosition-prop. I could just write it out instead of using the spread operator.

jenskuhrjorgensen commented on src/components/bottomSheetBackdrop/constants.ts:
```diff
@@ -4,10 +4,19 @@ const DEFAULT_DISAPPEARS_ON_INDEX = 0;
+const DEFAULT_ACCESSIBLE = true;
+const DEFAULT_ACCESSIBILITY_ROLE = 'button';
+const DEFAULT_ACCESSIBILITY_LABEL = 'Bottom Sheet backdrop';
+const DEFAULT_ACCESSIBILITY_HINT = 'Tap to close the Bottom Sheet';
```
```suggestion
const DEFAULT_ACCESSIBILITY_HINT = 'Tap to close the bottom sheet';
```

Mahmoud-SK replied:
Done.

jenskuhrjorgensen commented on src/components/bottomSheetBackdrop/constants.ts:
```diff
@@ -4,10 +4,19 @@ const DEFAULT_DISAPPEARS_ON_INDEX = 0;
 
+const DEFAULT_ACCESSIBLE = true;
+const DEFAULT_ACCESSIBILITY_ROLE = 'button';
+const DEFAULT_ACCESSIBILITY_LABEL = 'Bottom Sheet backdrop';
```
```suggestion
const DEFAULT_ACCESSIBILITY_LABEL = 'Bottom sheet backdrop';
```

jenskuhrjorgensen commented on src/components/bottomSheet/constants.ts:
```diff
@@ -33,6 +33,15 @@ const INITIAL_CONTAINER_OFFSET = {
 
+// accessibility
+const DEFAULT_ACCESSIBLE = true;
+const DEFAULT_ACCESSIBILITY_LABEL = 'Bottom Sheet';
```
```suggestion
const DEFAULT_ACCESSIBILITY_LABEL = 'Bottom sheet';
```

jenskuhrjorgensen commented on src/components/bottomSheetHandle/BottomSheetHandle.tsx:
```diff
@@ -2,11 +2,22 @@ import React, { memo, useMemo } from 'react';
+  accessibilityRole = DEFAULT_ACCESSIBILITY_ROLE,
+  accessibilityLabel = DEFAULT_ACCESSIBILITY_LABEL,
+  accessibilityHint = DEFAULT_ACCESSIBILITY_HINT,
+  ...rest
```
Why is the spread of the `rest` props needed?

jenskuhrjorgensen commented on src/components/bottomSheetHandle/constants.ts:
```diff
@@ -0,0 +1,12 @@
+const DEFAULT_ACCESSIBLE = true;
+const DEFAULT_ACCESSIBILITY_ROLE = 'adjustable';
+const DEFAULT_ACCESSIBILITY_LABEL = 'Bottom Sheet handle';
```
```suggestion
const DEFAULT_ACCESSIBILITY_LABEL = 'Bottom sheet handle';
```

jenskuhrjorgensen commented on src/components/bottomSheetHandle/constants.ts:
```diff
@@ -0,0 +1,12 @@
+const DEFAULT_ACCESSIBILITY_ROLE = 'adjustable';
+const DEFAULT_ACCESSIBILITY_LABEL = 'Bottom Sheet handle';
+const DEFAULT_ACCESSIBILITY_HINT =
+  'Drag up or down to extend or minimize the Bottom Sheet';
```
```suggestion
  'Drag up or down to extend or minimize the bottom sheet';
```

jenskuhrjorgensen reviewed (commented):
Hi @Mahmoud-SK 

I've added some review comments to your PR in order to maybe make the change a little smaller in an attempt to maybe get this merged/prioritized. 

Let me know what you think.

Best regards
Jens

jenskuhrjorgensen commented on src/components/bottomSheetBackdrop/BottomSheetBackdrop.tsx:
```diff
@@ -33,6 +37,11 @@ const BottomSheetBackdropComponent = ({
+  accessible: _providedAccessible = DEFAULT_ACCESSIBLE,
+  accessibilityRole: _providedAccessibilityRole = DEFAULT_ACCESSIBILITY_ROLE,
+  accessibilityLabel: _providedAccessibilityLabel = DEFAULT_ACCESSIBILITY_LABEL,
+  accessibilityHint: _providedAccessiblityHint = DEFAULT_ACCESSIBILITY_HINT,
```
```suggestion
  accessibilityHint: _providedAccessibilityHint = DEFAULT_ACCESSIBILITY_HINT,
```

jenskuhrjorgensen commented on src/components/bottomSheetBackdrop/BottomSheetBackdrop.tsx:
```diff
@@ -119,12 +128,11 @@ const BottomSheetBackdropComponent = ({
+        {...rest}
       >
         {children}
       </Animated.View>
```
I think you also need to add the accessibility props to the `<Animated.View>` below:
```
    <Animated.View
      style={containerStyle}
      pointerEvents={pointerEvents}
      accessible={_providedAccessible ?? undefined}
      accessibilityRole={_providedAccessibilityRole ?? undefined}
      accessibilityLabel={_providedAccessibilityLabel ?? undefined}
      accessibilityHint={_providedAccessibilityHint ?? undefined}
      {...rest}
    >
      {children}
    </Animated.View>
```
Maybe add this to a local const and use that because it is the exact same component as is used inside `<TapGestureHandler/>` when `pressBehavior !== 'none'`.

Mahmoud-SK replied:
Good idea, done.

Mahmoud-SK committed:
Minor adjustments after review

Mahmoud-SK committed:
removed unused spread operator

gorhom commented on src/components/bottomSheet/BottomSheet.tsx:
```diff
@@ -602,8 +619,42 @@ const BottomSheetComponent = forwardRef<BottomSheet, BottomSheetProps>(
           _providedOnChange(index);
         }
+
+        AccessibilityInfo.isScreenReaderEnabled().then(isEnabled => {
```
lets keep this logic out of the library and let user customise it as they pleased with `onChange` prop.

Mahmoud-SK replied:
I've removed the "AccessibilityInfo.isScreenReaderEnabled"-logic out of the library.

gorhom commented on src/components/bottomSheetBackdrop/BottomSheetBackdrop.tsx:
```diff
@@ -33,6 +37,11 @@ const BottomSheetBackdropComponent = ({
+  accessibilityRole: _providedAccessibilityRole = DEFAULT_ACCESSIBILITY_ROLE,
+  accessibilityLabel: _providedAccessibilityLabel = DEFAULT_ACCESSIBILITY_LABEL,
+  accessibilityHint: _providedAccessibilityHint = DEFAULT_ACCESSIBILITY_HINT,
+  ...rest
```
please remove `...rest`

Mahmoud-SK replied:
Removed the ...rest

gorhom commented on src/components/bottomSheetHandle/BottomSheetHandle.tsx:
```diff
@@ -2,11 +2,22 @@ import React, { memo, useMemo } from 'react';
+  accessibilityRole = DEFAULT_ACCESSIBILITY_ROLE,
+  accessibilityLabel = DEFAULT_ACCESSIBILITY_LABEL,
+  accessibilityHint = DEFAULT_ACCESSIBILITY_HINT,
+  ...rest
```
please remove `...rest`

Mahmoud-SK replied:
Removed the ...rest

gorhom reviewed (changes requested):
thanks @Mahmoud-SK, for submitting this pr, i have left some comments. and most importantly i think we should not introduce the change announcement, since i believe it is on the user to set it up not the library.

Mahmoud-SK committed:
fix: removed change announcement

removed the change announcement part of the code, and the ..rest from the two mentioned files.

aweary commented on src/components/bottomSheetBackdrop/BottomSheetBackdrop.tsx:
```diff
@@ -112,27 +120,35 @@ const BottomSheetBackdropComponent = ({
-        accessibilityHint={`Tap to ${
-          typeof pressBehavior === 'string' ? pressBehavior : 'move'
-        } the Bottom Sheet`}
+        accessible={_providedAccessible ?? undefined}
```
Why would you mark the root `View` as `accessible` for the bottom sheet? That makes it so you cannot focus any of the descendant views (which is all of the sheet content) with VoiceOver on iOS.

haileyok replied:
Yep, just tested this and this is the problem I'm seeing as well. @flo-sch I see you just opened an accessibility PR, did you want to take a look at this as well there?

aweary reviewed (commented):
@gorhom we recently updated our fork of this library and this commit broke VoiceOver for all of our sheets. I don't think this change makes sense for any sheet that would have interactive content in it (which I imagine is most?).

I would recommend reverting, this doesn't improve accessibility at all.
